### PR TITLE
Docs: Add topic section on storage

### PR DESCRIPTION
**Pull request by sphuber**
Gives overview of the storage plugins that ship with `aiida-core` with their strengths and weaknesses. This can be referred to from other parts of the docs when recommending which storage to choose.

### CODE REVIEW

File: docs/source/topics/storage.rst [hunk 1]
```diff
@@ -0,0 +1,89 @@
+.. _topics:storage:
+
+*******
+Storage
+*******
+
+Each AiiDA profile defines a _storage_ which is where all data in the provenance graph is stored.
```
Comment (mbercx) [inline docs/source/topics/storage.rst:7]:
```suggestion
Each AiiDA profile defines a *storage* which is where all data in the provenance graph is stored.
```

It's clearly time for us to move to a Markdown-based documentation approach. ^^

File: docs/source/topics/storage.rst [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+      :fa:`plus;mr-1` **Strengths**:
+
+      - Easy to setup
+      - Requires no running services
+
+      :fa:`minus;mr-1` **Weaknesses**:
+
+      - Some ``QueryBuilder`` functionality is not supported
+      - Performance of SQLite is inferior to PostgreSQL
```
Comment (mbercx) [inline docs/source/topics/storage.rst:53]: Just for my understanding: what do we mean by "performance" here? It's slower? Or can things go awry?

File: docs/source/topics/storage.rst [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+      :fa:`plus;mr-1` **Strengths**:
+
+      - Easy to setup
+      - Requires no running services
+
+      :fa:`minus;mr-1` **Weaknesses**:
+
+      - Some ``QueryBuilder`` functionality is not supported
+      - Storage is deleted once session is closed
```
Comment (mbercx) [inline docs/source/topics/storage.rst:71]: Is this true? Also: isn't this storage backend read-only once created?

File: docs/source/topics/storage.rst [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+   .. grid-item-card:: ``core.sqlite_temp``
+
+      *Temporary storage mostly used for unit testing.*
```
Comment (mbercx) [inline docs/source/topics/storage.rst:75]:
```suggestion
      *Temporary in-memory storage, mostly used for unit testing or demonstrations.*
```

File: docs/source/topics/storage.rst [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+      :fa:`database;mr-1` SQLite database
+
+      :fa:`file;mr-1` ``disk-objectstore`` container
+
+      :fa:`plus;mr-1` **Strengths**:
+
+      - Easy to setup
+      - Requires no running services
```
Comment (mbercx) [inline docs/source/topics/storage.rst:84]: I'd also say a strength here is automatic cleanup. :)

File: docs/source/topics/storage.rst [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      *Storage contained in single ZIP file used for export archives.*
+
+      :fa:`database;mr-1` SQLite database
+
+      :fa:`file;mr-1` ``disk-objectstore`` container
+
+      :fa:`plus;mr-1` **Strengths**:
+
+      - Easy to setup
+      - Requires no running services
```
Comment (mbercx) [inline docs/source/topics/storage.rst:66]: A strength here is that it's easy to backup and transfer the data.

File: docs/source/topics/storage.rst [hunk 7]
```diff
@@ -0,0 +1,89 @@
+.. _topics:storage:
+
+*******
+Storage
+*******
+
+Each AiiDA profile defines a _storage_ which is where all data in the provenance graph is stored.
+Typically, a storage consists of a :ref:`database <topics:database>` and a :ref:`repository <topics:repository>`.
+The provenance graph data itself is mostly persisted to the database, with binary blobs that are attached to nodes, such as files, being stored in the repository.
```
[the remark this answers is not in the document]
Comment (sphuber) [reply docs/source/topics/storage.rst:9]: I mean that technically what you store don't _have_ to be files. They can literally be some binary content, e.g.
```
node.base.repository.put_object_from_filelike('obj', io.BytesIO(b'some content'))
```
I tried to keep the language generic, but if it is confusing, we can simplify it.

File: docs/source/topics/storage.rst (hunk 2 above)
Comment (sphuber) [reply docs/source/topics/storage.rst:53]: Performance is performance, i.e., it is expected to be slower during operation, especially for larger databases and when it is accessed by multiple processes concurrently, i.e. multiple daemon workers, interactive shells.

File: docs/source/topics/storage.rst [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+      :fa:`minus;mr-1` **Weaknesses**:
+
+      - Requires a service running (PostgreSQL)
+
+   .. grid-item-card:: ``core.sqlite_dos``
+
+      *Easy to setup storage for tests, demos and experimenting.*
+
+      :fa:`database;mr-1` SQLite database
```
[the remark this answers is not in the document]
Comment (sphuber) [reply docs/source/topics/storage.rst:41]: In the `filepath` config storage option. User can define the directory on disk, with it defaulting to the systems tmp file system if not provided

File: docs/source/topics/storage.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbercx) [reply docs/source/topics/storage.rst:9]: Perhaps

```suggestion
The provenance graph data itself is mostly persisted to the database, whereas files attached to nodes (or other binary content) is stored in the repository.
```

File: docs/source/topics/storage.rst (hunk 6 above)
Comment (sphuber) [reply docs/source/topics/storage.rst:66]: Yes, although with the upcoming `verdi storage backup` the same could be said for `core.psql_dos`.

File: docs/source/topics/storage.rst (hunk 3 above)
Comment (sphuber) [reply docs/source/topics/storage.rst:71]: No this is incorrect. And yes, it should be read-only. Will add that.

**Review by mbercx — APPROVED** [review]
Thanks @sphuber, no more notes!

**Pull request merged**